src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs

```diff
@@ -75,47 +81,52 @@ private async Task<string> GetRegionAsync(RequestContext requestContext)
 
                 // A bad request occurs when the version in the IMDS call is no longer supported.
                 if (response.StatusCode == HttpStatusCode.BadRequest)
                 {
-                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers).ConfigureAwait(false); // Get the latest version
-                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger).ConfigureAwait(false); // Call again with updated version
+                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers, requestContext.UserCancellationToken).ConfigureAwait(false); // Get the latest version
+                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger, retry: false, GetCancellationToken(requestContext.UserCancellationToken)).ConfigureAwait(false); // Call again with updated version
                 }
 
                 if (response.StatusCode == HttpStatusCode.OK && !response.Body.IsNullOrEmpty())
                 {
-                    LocalImdsResponse localImdsResponse = JsonHelper.DeserializeFromJson<LocalImdsResponse>(response.Body);
-
-                    if (localImdsResponse != null && !localImdsResponse.location.IsNullOrEmpty())
-                    {
-                        logger.Info($"[Region discovery] Call to local IMDS returned region: {localImdsResponse.location}");
-                        LogTelemetryData(localImdsResponse.location, RegionSource.Imds, requestContext);
-
-                        return localImdsResponse.location;
-                    }
+                    return response.Body;
                 }
                     
                 logger.Info($"[Region discovery] Call to local IMDS failed with status code: {response.StatusCode} or an empty response.");
 
                 throw MsalServiceExceptionFactory.FromImdsResponse(
-                MsalError.RegionDiscoveryFailed,
-                MsalErrorMessage.RegionDiscoveryFailed,
-                response);
-            }
-            catch (MsalServiceException)
+                    MsalError.RegionDiscoveryFailed,
+                    MsalErrorMessage.RegionDiscoveryFailed,
+                    response);
+            }  
+            catch (MsalServiceException e)
             {
-                throw;
+                if (MsalError.RequestTimeout.Equals(e.ErrorCode))
+                {
+                    throw new MsalServiceException(MsalError.RegionDiscoveryFailed, MsalErrorMessage.RegionDiscoveryFailedWithTimeout, e);
+                }
+
+                throw e;
             }
             catch (Exception e)
             {
-                logger.Info("[Region discovery] Call to local imds failed." + e.Message);
-                throw new MsalServiceException(MsalError.RegionDiscoveryFailed, MsalErrorMessage.RegionDiscoveryFailed);
+                logger.Info("[Region discovery] Call to local imds failed. " + e);
```

Comment:
Actually this should go through the Pii logs, as the exception could have pii, as we don't know what's in it.

Reply:
This exception would mostly be either timeout exception or any other exception where we were unable to get a response from imds. There is no Pii data involved here.